### PR TITLE
CRM-20621 : manage tags: the tag usage count is not accurate

### PR DESCRIPTION
**Pull request by monishdeb**
* [CRM-20621: manage tags: the tag usage count is not accurate](https://issues.civicrm.org/jira/browse/CRM-20621)

### CODE REVIEW

**Comment by lcdservices** [comment]
looks good. count is accurate and query is compatible with mysql 5.7

**Comment by colemanw** [comment]
@monishdeb can you please fix the failing test?

**Comment by colemanw** [comment]
@civicrm-builder retest this please

**Comment by colemanw** [comment]
@monishdeb and @lcdservices I know it's been a year since this was merged but I'm hitting a major performance issue with it on my sandbox. Clicking on a tagset with a few dozen tags in it takes 8 seconds to load, I can imagine it would take significantly longer on a live site with a few hundred tags in a set.
The major bottleneck appears to be the `'usages' => civicrm_api3('EntityTag', 'getcount'...` for every item in the list. That's a lot of api calls! And commenting it out takes the load time to < 1 second.
But reading the code, it's also now calling up `$childTagIDs = CRM_Core_BAO_Tag::getChildTags();` which is building a potentially massive array in memory, only to discard most of it which is redundant with the query above.

File: CRM/Admin/Page/AJAX.php [hunk 1]
```diff
@@ -304,20 +304,24 @@ public static function getTagTree() {
...
-      ORDER BY tag.name";
+    $parentClause = $parent ? "AND parent_id = $parent" : 'AND parent_id IS NULL';
+    $sql = "SELECT *
+      FROM civicrm_tag
+      WHERE is_tagset <> 1 $parentClause
+      GROUP BY id
+      ORDER BY name";
+
+    // fetch all child tags in Array('parent_tag' => array('child_tag_1', 'child_tag_2', ...)) format
+    $childTagIDs = CRM_Core_BAO_Tag::getChildTags();
```
Comment (colemanw) [inline CRM/Admin/Page/AJAX.php]: Is this absolutely necessary? If we don't need the usage data can we re-add the `LEFT JOIN civicrm_tag child ON child.parent_id = tag.id` back to the query? That seems to be the only purpose of loading this massive array into memory.

File: CRM/Admin/Page/AJAX.php [hunk 2]
```diff
@@ -330,18 +334,25 @@ public static function getTagTree() {
...
+            'tag_id' => $dao->id,
+          )),
         ),
       );
     }
 
+    if (!empty($_REQUEST['is_unit_test'])) {
+      return $result;
+    }
+
```
Comment (colemanw) [inline CRM/Admin/Page/AJAX.php]: We could remove this hack by making the `CRM_Admin_Page_AJAX::getTagTree()` function a one-line wrapper and move the logic to a BAO function that would do the work and could be more elegantly unit-tested.

**Comment by lcdservices** [comment]
@colemanw FWIW we're using this on NYSS where some sites have ~2000 tags, and we have not noticed a performance hit on that page. I'll do some more testing to confirm, but it's been in production with this change for quite some time without complaint.

File: CRM/Admin/Page/AJAX.php (hunk 1 above)
Comment (monishdeb) [reply CRM/Admin/Page/AJAX.php]: Problem with the left join is that it doesn't look deep down the hierarchy to retrieve level-N child tags, reason why I need to introduce this new fn to fetch the tag tree via recursive search. However, we can avoid this lookup on each call by storing the  ```$childTagIDs``` array in a ```Civi::static``` variable? And need to ensure that whenever a tag got deleted/moved in the tag tree this static variable got repopulated.

Comment (colemanw) [reply CRM/Admin/Page/AJAX.php]: This function, by design, does not need to look deep within the hierarchy. It only needs to know whether the tag has children or not. 1 level deep is fine.

**Pull request merged**